crates/ark/src/data_explorer/format.rs

```diff
@@ -49,7 +49,33 @@ pub fn format_string(x: SEXP, format_options: &FormatOptions) -> Vec<String> {
 }
 
 fn format(x: SEXP, format_options: &FormatOptions) -> Vec<FormattedValue> {
-    format_values(x, format_options).unwrap_or(unknown_format(x))
+    let mut formatted = format_values(x, format_options).unwrap_or(unknown_format(x));
+
+    // Truncate the values if they are too long
+    formatted.iter_mut().for_each(|v| {
+        if let FormattedValue::Value(x) = v {
+            truncate_inplace(x, format_options.max_value_length);
+        }
+    });
+
+    formatted
+}
+
+// Truncating strings in Rust is more complicated that one would imagine.
+// If you index using eg s[..6] that would take 6 bytes, which is not necessarily 6 characters.
+// We need to iterate over the characters and truncate the string at the right character.
+// See also https://doc.rust-lang.org/book/ch08-02-strings.html
+fn truncate_inplace(s: &mut String, max_len: i64) {
+    // Returns the indice of the nth character in the string if it exists
+    let index = s.char_indices().nth(max_len as usize);
+
+    // If an index is found, truncate the string to that index
+    match index {
+        // truncate is an in-place operation that takes the new max index as input
+        // but looking the string as a sequence of bytes, not chars.
+        Some((i, _)) => s.truncate(i),
```

Comment:
Is this right? Doesn't `i` correspond to the _i-th character_? But `truncate()` is looking for _the i-th byte_?

Reply:
Ah according to this, the `i` there is the _byte index_ of that n-th character, so I think you are right
https://stackoverflow.com/questions/38461429/how-can-i-truncate-a-string-to-have-at-most-n-characters

Reply:
Yep, it's the byte index for the specific character: https://doc.rust-lang.org/std/primitive.str.html#method.char_indices

Reply:
👍  i was looking for information about what the first element in the tuple corresponded to over on the `CharIndices` iterator page, but couldn't find it! Didn't think to look at the str method page
https://doc.rust-lang.org/std/str/struct.CharIndices.html